cmd/object-handlers.go

```diff
@@ -534,13 +535,55 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	// Copy source object to destination, if source and destination
-	// object is same then only metadata is updated.
-	objInfo, err := objectAPI.CopyObject(ctx, srcBucket, srcObject, dstBucket, dstObject, srcInfo)
-	if err != nil {
-		pipeWriter.CloseWithError(err)
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
+	var objInfo ObjectInfo
+
+	_, err = objectAPI.GetBucketInfo(ctx, dstBucket)
+	if err == toObjectErr(errVolumeNotFound, dstBucket) && !cpSrcDstSame {
+		if globalDNSConfig != nil {
+			if dstRecord, errEtcd := globalDNSConfig.Get(dstBucket); errEtcd == nil {
+				go func() {
+					if err = objectAPI.GetObject(ctx, srcBucket, srcObject, 0, srcInfo.Size, srcInfo.Writer, srcInfo.ETag); err != nil {
+						pipeWriter.CloseWithError(err)
+						writeErrorResponse(w, ErrInternalError, r.URL)
+						return
+					}
+					// Close writer explicitly to indicate data has been written
+					defer srcInfo.Writer.Close()
+				}()
+				// Forward request to appropriate instance
+				client, err := getRemoteInstanceClient(dstRecord)
+				if err != nil {
+					pipeWriter.CloseWithError(err)
+					writeErrorResponse(w, ErrInternalError, r.URL)
+					return
+				}
+				if _, err = client.PutObject(dstBucket, dstObject, srcInfo.Reader, srcInfo.Size, "", "", srcInfo.UserDefined); err != nil {
+					pipeWriter.CloseWithError(err)
+					writeErrorResponse(w, ErrInternalError, r.URL)
+					return
+				}
+				remoteObjInfo, err := client.StatObject(dstBucket, dstObject, miniogo.StatObjectOptions{})
```

Comment:
why do you need a statObject? putObject already returns the objectInfo in the previous call.